trainer: migration guide from training operator to Trainer V2

akiseakusa opened the pull request:
<!-- Add the component name to the PR's title. Example: pipelines: Fixed broken link in Getting Started with Kubeflow Pipelines -->


**Checklist:**
- [x] You have signed off your commits! ([sign-off guide](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits))
- [x] Ensure you follow best practices from our guide. [Contributing]https://github.com/kubeflow/website/blob/master/content/en/docs/about/contributing.md)
- [ ] You have included screenshots when changing the website style or adding a new page.


**Description of your changes:**

This PR adds a new migration guide under the Kubeflow Trainer documentation.

It provides users with a clear migration path from the legacy Training Operator (v1) APIs (e.g., PyTorchJob) to the new unified Trainer V2 APIs (e.g., TrainingRuntime). The guide includes:

- A side-by-side example of migrating a PyTorchJob to TrainingRuntime.
- A compatibility table for Trainer versions, CRDs, and Kubernetes requirements.
- Notes on key changes, deprecations, and architectural improvements in Trainer V2.

This contribution addresses the request in [kubeflow/trainer#2412 [(comment)]](https://github.com/kubeflow/trainer/issues/2412#issuecomment-3054291878)

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
```diff
@@ -23,6 +23,89 @@ The key features introduced by Kubeflow Trainer are:
   MPI performance.
 
-## Migration Paths
+## Migration paths
```
We can keep section titles capitalised. 

```suggestion
## Migration Paths
```

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
```diff
@@ -23,6 +23,89 @@ The key features introduced by Kubeflow Trainer are:
+
+Kubeflow Trainer v2 introduces new APIs that replace the older, framework-specific CRDs such as `PyTorchJob`, `TFJob`, and `MPIJob`. These new APIs—`TrainingRuntime`, `TrainJob`, and `ClusterTrainingRuntime`—offer a more flexible and unified interface for defining training jobs across frameworks.
+
+This section shows how to migrate an existing PyTorch job to the new API.
```
```suggestion
This section shows how to migrate an existing PyTorchJob to the new API.
```

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
````diff
@@ -23,6 +23,89 @@ The key features introduced by Kubeflow Trainer are:
+              image: docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727
+              command: ["python3", "/opt/pytorch-mnist/mnist.py", "--epochs=1"]
+```
+### New: TrainingRuntime (v2alpha1)
+```yaml
+apiVersion: trainer.kubeflow.org/v2alpha1
+kind: TrainingRuntime
````
Let's use ClusterTrainingRuntime as a default migration path, since it is the default runtimes to use.
```suggestion
### New: ClusterTrainingRuntime (v2alpha1)

```yaml
apiVersion: trainer.kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
```

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
````diff
@@ -23,6 +23,89 @@ The key features introduced by Kubeflow Trainer are:
+                      value: "29400"
+                  command: ["torchrun", "train.py"]
+```
+The new `TrainingRuntime` CRD allows you to define reusable templates and use MPI-based distributed training with greater flexibility. The `replicatedJobs` and `mpiPolicy` fields replace framework-specific roles like `Master` and `Worker`.
````
MPI is used only for ML frameworks that require MPI functionality (e.g. DeepSpeed). So it is not needed for `torchrun`.

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
```diff
@@ -23,6 +23,89 @@ The key features introduced by Kubeflow Trainer are:
+
+The following table lists compatibility between Kubeflow Trainer versions, supported APIs, and minimum Kubernetes versions:
+
+| Trainer version | API version | CRDs introduced                                | Minimum Kubernetes version | Notes                        |
+|-----------------|-------------|-------------------------------------------------|-----------------------------|-------------------------------|
+| v1.4–v1.8       | v1          | `PyTorchJob`, `TFJob`, `MPIJob`, etc.          | 1.23+                       | Training Operator style       |
+| v1.9+           | v2alpha1    | `TrainingRuntime`, `TrainJob`, `ClusterTrainingRuntime` | 1.31.3+                     | Trainer v2 (breaking changes) |
```
This table is incorrect, since we don't support TrainJob in Training Operator 1.9.
I would suggest remove it.

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
```diff
@@ -23,6 +23,89 @@ The key features introduced by Kubeflow Trainer are:
+### Additional information
+
+- Trainer v2 does not use individual CRDs per framework. Instead, it abstracts functionality into a unified runtime specification.
+- You can use the `Kubeflow Python SDK` to generate and submit Trainer v2 workloads programmatically.
```
We should explain that Kubeflow Python SDK is the primary interface for ML Users to interact with TrainJob APIs.

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
```diff
@@ -23,6 +23,89 @@ The key features introduced by Kubeflow Trainer are:
+- Trainer v2 does not use individual CRDs per framework. Instead, it abstracts functionality into a unified runtime specification.
+- You can use the `Kubeflow Python SDK` to generate and submit Trainer v2 workloads programmatically.
+- Initializer and multi-node coordination logic is managed by the system, reducing overhead on the training container.
+- Support for specifying `managedBy` fields and runtime API versions (via `runtimeRef.version`) is under consideration.
```
What do you mean by that ?

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
```diff
@@ -23,6 +23,89 @@ The key features introduced by Kubeflow Trainer are:
+- Initializer and multi-node coordination logic is managed by the system, reducing overhead on the training container.
+- Support for specifying `managedBy` fields and runtime API versions (via `runtimeRef.version`) is under consideration.
+
+You can find detailed proposals and discussion [in the Kubeflow Trainer GitHub repository](https://github.com/kubeflow/trainer/tree/master/docs/proposals).
```
Please add link to the KEP: https://github.com/kubeflow/trainer/tree/master/docs/proposals/2170-kubeflow-trainer-v2

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
```diff
@@ -23,6 +23,89 @@ The key features introduced by Kubeflow Trainer are:
+
+This section shows how to migrate an existing PyTorch job to the new API.
+
+### Migrate PyTorchJob to TrainingRuntime
```
I think, this example should be converted differently.
We should use [the default Torch runtime](https://github.com/kubeflow/trainer/blob/master/manifests/base/runtimes/torch_distributed.yaml), and use TrainJob to override:
```
image:
command:
numNodes
```

You should explain that Kubeflow Trainer also provides Python SDK to allow ML Users submit TrainJob without `kubectl`.

akiseakusa replied:
I’ve updated the migration guide with a TrainJob configuration, using the default Torch runtime, where MPI isn’t needed for this torchrun-based example since it relies on PyTorch’s native backends (e.g., NCCL/Gloo). I’ve included replicas: 2 to override the number of nodes, as you suggested, and removed the mpiPolicy to simplify the setup. The updated configuration is as follows:

<pre> 
apiVersion: trainer.kubeflow.org/v2alpha1
kind: TrainJob
metadata:
  name: pytorch-simple
  namespace: kubeflow
spec:
  runtimeRef:
    name: torch-distributed 
  template:
    spec:
      containers:
        - name: trainer
          image: docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727
          command: ["torchrun", "/opt/pytorch-mnist/mnist.py", "--epochs=1"]
          env:
            - name: MASTER_ADDR
              value: "pytorch-simple-0"
            - name: MASTER_PORT
              value: "29400"
      replicas: 2
</pre>

I’ve also added a note explaining that Kubeflow Trainer provides a Python SDK, allowing ML users to submit TrainJob workloads programmatically without Kubectl, as recommended. Please let me know if you have further suggestion to improve.

andreyvelich replied:
You should use [.spec.trainer.numNodes API](https://github.com/kubeflow/trainer/blob/master/pkg/apis/trainer/v1alpha1/trainjob_types.go#L222) and remove the MASTER_ADDR, MASTER_PORT from env vars, since these values are set by controller automatically.

andreyvelich reviewed (commented):
Thanks @akiseakusa, I added a few comments.
/cc @astefanutti @Electronic-Waste @kubeflow/wg-training-leads @kramaranya

Arhell reviewed (commented):
/ok-to-test

akiseakusa commented:
Hi, I’ve updated the PR title to follow the required format (trainer: migration guide from training operator to Trainer V2) and adjusted the workflow in my fork to include the correct permissions for commenting. The PR title check is now awaiting maintainer approval to run. Could someone please review and approve the workflow run?

andreyvelich reviewed (commented):
@akiseakusa @kramaranya @kubeflow/wg-training-leads @tenzen-y @Electronic-Waste @astefanutti @franciscojavierarceo @varodrig I updated the migration guide, so the TrainJob example is working.

If that looks good, let's merge it, so we can add it to the Changelog and Blog post.

astefanutti commented:
Looks fantastic, thanks!

/lgtm

andreyvelich reviewed (commented):
Thanks for your contribution @akiseakusa!
/approve
